Sound module script

coolpeter98 opened the pull request:
This adds a module script which allows setting an alias for sound ids, and also adding multiple sounds under one alias, for more variety (exempli gratia varying crunching sounds when eating golden apple)

coolpeter98 committed:
Add sound module script and sound when placing blocks

coolpeter98 committed:
Make the concatenation Cooler.

EliTheGingerCat commented on src/ReplicatedStorage/client/cubes.client.luau:
```diff
@@ -65,10 +66,11 @@ local function on_input_began(input: InputObject): ()
 
-UserInputService.InputBegan:Connect(on_input_began)
\ No newline at end of file
+UserInputService.InputBegan:Connect(on_input_began)
```
Apparently it is good to have newlines at the end of text files so I guess I will start doing it...

Source: <https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline>

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,25 @@
+--[[
+
+placeholder sounds for now
+if anyone has some Cool sounds that can be used here, please submit a pull request :3
```
Putting that face in the source code is wild.

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,25 @@
+local function play_sound(sound)
+    local soundlist = sounds[sound]
+    local soundid = soundlist and soundlist[math.random(1, #soundlist)]
+if not soundid then
+    warn("Invalid sound entered")
+end
```
The indentation does not fit with the rest of the function.

coolpeter98 replied:
I just removed it altogether

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,25 @@
+
+local function play_sound(sound)
+    local soundlist = sounds[sound]
+    local soundid = soundlist and soundlist[math.random(1, #soundlist)]
```
Why would `soundlist` be nil? I agree it can be nillable since there is no typechecking here so I would just add an assertion. I would assert `soundid` too since the table can be empty. It is NOT okay for `soundlist` or `soundid` to be nil.

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,25 @@
+if not soundid then
+    warn("Invalid sound entered")
+end
+    local instance = Instance.new("Sound")
+    instance.Parent = game:WaitForChild("SoundService")
+    instance.SoundId = `rbxassetid://{soundid}`
+    instance:Play()
```
This seems to be a memory leak. More and more `Sound` instances will be created over time. I think it might be better to use [`Sound.PlayOnRemove`](https://create.roblox.com/docs/reference/engine/classes/Sound#PlayOnRemove) and destroy the instance after setting `SoundId`. Also, this function ought to take a position parameter so that the sound can be played in the right spot.

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,25 @@
+local sounds = {
+    block_place = {2164165859, 6586979979}, --multiple sounds can be added for variety
```
This is creative, I like it!

EliTheGingerCat reviewed (changes requested):
I am going to add my requested changes.

coolpeter98 committed:
Make comment more serious (literally 1984..)

coolpeter98 committed:
Fix indentation

coolpeter98 committed:
Fix memory leak

coolpeter98 committed:
Add position option to play_sound, move sound playing to server so other players can hear the blocks.
Would be good if the blocks placed by the local player still had no delay, though I'm not sure how i could do it without making duplicate sounds.

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,33 @@
+    part.Parent = game:WaitForChild("Workspace")
+    sound.Parent = part
+
+    print(position)
```
Debug print should be removed.

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,33 @@
+    local part = Instance.new("Part")
+    local sound = Instance.new("Sound")
+
+    part.Parent = game:WaitForChild("Workspace")
```
Why wait for Workspace... There is a [`workspace`](https://create.roblox.com/docs/reference/engine/globals/RobloxGlobals#workspace) global for a reason. Though I am considering not using it anymore since it is inconsistent with every other service. Regardless, one does not wait for services, one [`gets`](https://create.roblox.com/docs/reference/engine/classes/ServiceProvider#GetService) them.

EliTheGingerCat commented on src/ReplicatedStorage/modules/play_sound.luau:
```diff
@@ -0,0 +1,33 @@
+    assert(soundlist, "Sound list cannot be nil.")
+    assert(soundid, "Sound ID cannot be nil.")
+
+    local part = Instance.new("Part")
```
The part should be anchored, have no collision, et cetera.

coolpeter98 replied:
Why? It is immediately destroyed before a frame is rendered (I think)

EliTheGingerCat replied:
Fair enough. Let us leave it for now and change it if it causes problems later.

EliTheGingerCat reviewed (changes requested):
Once again, I am willing to add these changes myself.

EliTheGingerCat commented:
By the way, do we have permission to use these sounds?

EliTheGingerCat committed:
style: add function types

EliTheGingerCat committed:
style: rename variables

EliTheGingerCat committed:
fix: improve error messages

EliTheGingerCat committed:
fix: remove debug print

EliTheGingerCat committed:
refactor: reorder statements

coolpeter98 commented:
> By the way, do we have permission to use these sounds?

probably, I just took them from the roblox toolbox, would be dumb if we *weren't* allowed to use them

EliTheGingerCat committed:
feat: play sound immediately for block placer

EliTheGingerCat committed:
fix: remove unused sound category

EliTheGingerCat committed:
docs: add credits
